main/src/ca/uwaterloo/flix/language/phase/EnumGen.scala

```diff
@@ -0,0 +1,650 @@
+/*
+ * Copyright 2015-2017 Ramin Zarifi
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ca.uwaterloo.flix.language.phase
+
+import ca.uwaterloo.flix.api.{Flix, TagInterface}
+import ca.uwaterloo.flix.language.CompilationError
+import ca.uwaterloo.flix.language.ast.ExecutableAst.Definition
+import ca.uwaterloo.flix.language.ast.Symbol.EnumSym
+import org.objectweb.asm
+import org.objectweb.asm.{ClassWriter, Label}
+import org.objectweb.asm.Opcodes._
+import ca.uwaterloo.flix.language.ast.{Type, _}
+import ca.uwaterloo.flix.util.{Evaluation, InternalCompilerException}
+import ca.uwaterloo.flix.util.Validation
+import ca.uwaterloo.flix.util.Validation._
+import CodegenHelper._
+
+object EnumGen extends Phase[ExecutableAst.Root, ExecutableAst.Root] {
+
+  /**
+    * At this phase, we create java classes and interfaces representing Enum types and Enum cases.
+    * The steps that we take to generate enums is as follows:
+    *
+    * 1. Transform non-functions and generate interface names
+    * We transform non functions to functions with no argument, then we use this list of definitions
+    * to generate interface names which is required in further steps to emit the correct signature for enum functions
+    *
+    * 2. Find enums and group them by symbols.
+    * We will recursively traverse the AST to find all enum case types. We will also extract the type of the field of
+    * the enum at this stage, so each element of the list is (EnumType, (caseName, fieldType)).
+    * Then we will group enums based on their enum symbols.
+    *
+    * 3. Generate byteCodes of Enum Interface.
+    * At this stage, we generate an interface for each enum symbol. All cases of that symbol will implement this interface.
+    * This interface extends `tagInterface` interface which has two methods: `getBoxedValue()` and `getTag()`. For example,
+    * for enum result defined as follows:
+    *
+    * enum Result[t, e] {
+    *   case Ok(t),
+    *   case Err(e)
+    * }
+    *
+    * we generate the bytecode corresponding to following class:
+    *
+    * package ca.waterloo.flix.enums.Result;
+    * import ca.uwaterloo.flix.api.TagInterface;
+    * public interface EnumInterface extends TagInterface {
```

Comment:
No it's called EnumInterface. My reasoning was to make Result be the path that contains the cases and the interface so virtually, in the folder Result we have TagInterface, Ok and Err